### PR TITLE
log execution times for analyses and remove log.Label usages in staticanalyze.go

### LINKED ISSUE #605: Record analysis runtimes
Currently there is no record of how long dynamic or static analysis takes. In order to have an idea of performance, it would be good to record the duration of each analysis, either by logging the time taken or including it into the analysis results (or both)

### PR DESCRIPTION
**Pull request by maxfisher-g**
Prints execution times for dynamic and static analysis to logs

Fixes #605 

Signed-off-by: Max Fisher <maxfisher@google.com>

### CODE REVIEW

File: internal/worker/rundynamic.go [hunk 1]
```diff
@@ -50,20 +53,27 @@ func RunDynamicAnalysis(pkg *pkgecosystem.Pkg, sbOpts []sandbox.Option) (result.
...
 
+		runDuration := time.Since(startTime)
+		durationString := fmt.Sprintf("%.1fs", runDuration.Seconds())
+
 		if err != nil {
 			// Error when trying to actually run; don't record the result for this phase
 			// or attempt subsequent phases
 			lastStatus = ""
 			lastError = err
+			log.Warn("Analysis failed after "+durationString, "phase", phase)
```
Comment (calebbrown) [inline internal/worker/rundynamic.go:68]: `runDuration` should be included at the end with a key (`..., "duration", runDuration`) so it is available in the structured logging.

`durationString` in the message itself is optional.

(Same goes for the other logging calls as well)

Comment (calebbrown) [reply internal/worker/rundynamic.go:68]: Also these log messages may need to have ecosystem, package and version included.

**Review by calebbrown — CHANGES_REQUESTED** [review]

File: internal/worker/rundynamic.go (hunk 1 above)
Comment (maxfisher-g) [reply internal/worker/rundynamic.go]: Done

File: internal/worker/rundynamic.go [hunk 2]
```diff
@@ -50,9 +54,20 @@ func RunDynamicAnalysis(pkg *pkgecosystem.Pkg, sbOpts []sandbox.Option) (result.
...
 		lastRunPhase = phase
 
+		runDuration := time.Since(startTime)
+		log.Info("Dynamic analysis phase finished",
+			log.Label("ecosystem", pkg.EcosystemName()),
+			log.Label("name", pkg.Name()),
+			log.Label("version", pkg.Version()),
+			log.Label("phase", string(phase)),
+			log.Label("error", strconv.FormatBool(err != nil)),
+			log.Label("dynamic_analysis_phase_duration_sec", fmt.Sprintf("%.1f", runDuration.Seconds())),
```
Comment (calebbrown) [inline internal/worker/rundynamic.go:68]: This probably should be a regular key, value field, not wrapped in `log.Label`.

i.e. `"phase_duration", runDuration`

The docs aren't totally clear, but `log.Label` is used to add labels to the GCP's logging. There are cardinality limits as per https://cloud.google.com/logging/docs/logs-based-metrics/labels.

In fact, thinking about this now, our use of log.Label hasn't been particularly careful in the past :grimacing:

**Review by calebbrown — CHANGES_REQUESTED** [review]

File: internal/worker/rundynamic.go [hunk 3]
```diff
@@ -50,9 +54,20 @@ func RunDynamicAnalysis(pkg *pkgecosystem.Pkg, sbOpts []sandbox.Option) (result.
...
+		phaseResult, err := dynamicanalysis.Run(sb, pkg.Command(phase))
 		lastRunPhase = phase
 
+		runDuration := time.Since(startTime)
+		log.Info("Dynamic analysis phase finished",
+			log.Label("ecosystem", pkg.EcosystemName()),
+			log.Label("name", pkg.Name()),
+			log.Label("version", pkg.Version()),
+			log.Label("phase", string(phase)),
+			log.Label("error", strconv.FormatBool(err != nil)),
```
Comment (calebbrown) [inline internal/worker/rundynamic.go:67]: Is "error" is probably better named "is_error" or "returned_error"

**Review by calebbrown — CHANGES_REQUESTED** [review]

File: internal/worker/rundynamic.go (hunk 3 above)
Comment (maxfisher-g) [reply internal/worker/rundynamic.go:67]: I decided to just log the actual error

File: internal/worker/rundynamic.go (hunk 2 above)
Comment (maxfisher-g) [reply internal/worker/rundynamic.go:68]: I've removed all the log.Label usages

File: sandboxes/staticanalysis/staticanalyze.go [hunk 4]
```diff
@@ -199,20 +202,20 @@ func run() (err error) {
...
-		log.Label("version", *version),
-		log.Label("local_path", *localFile),
-		log.Label("output_file", *output),
-		log.Label("analyses", strings.Join(uniqueAnalyses, ",")))
+		"ecosystem", *ecosystem,
+		"package", *packageName,
+		"version", *version,
+		"local_path", *localFile,
+		"output_file", *output,
+		"analyses", strings.Join(uniqueAnalyses, ","))
```
Comment (calebbrown) [inline sandboxes/staticanalysis/staticanalyze.go:210]: You don't need to string-ify this, zap should convert this into a JSON array.

File: internal/worker/rundynamic.go [hunk 5]
```diff
@@ -50,9 +53,20 @@ func RunDynamicAnalysis(pkg *pkgecosystem.Pkg, sbOpts []sandbox.Option) (result.
...
 	var lastError error
 	for _, phase := range pkg.Manager().RunPhases() {
-		result, err := dynamicanalysis.Run(sb, pkg.Command(phase))
+		startTime := time.Now()
+		phaseResult, err := dynamicanalysis.Run(sb, pkg.Command(phase))
 		lastRunPhase = phase
 
+		runDuration := time.Since(startTime)
+		log.Info("Dynamic analysis phase finished",
+			"ecosystem", pkg.EcosystemName(),
```
Comment (calebbrown) [inline internal/worker/rundynamic.go:62]: `ecosystem` and `phase` and `status` can all be wrapped in `log.Label()` as they have fixed cardinality.

File: internal/worker/runstatic.go [hunk 6]
```diff
@@ -77,7 +80,14 @@ func RunStaticAnalyses(pkg *pkgecosystem.Pkg, sbOpts []sandbox.Option, tasks ...
 
 	status := analysis.StatusForRunResult(runResult)
 
-	// TODO log status
+	totalTime := time.Since(startTime)
+	log.Info("Static analysis finished",
+		"ecosystem", pkg.EcosystemName(),
+		"name", pkg.Name(),
+		"version", pkg.Version(),
+		"status", string(status),
```
Comment (calebbrown) [inline internal/worker/runstatic.go:88]: The name/key `"status"` is ambiguous and probably should include some more context

**Review by calebbrown — APPROVED** [review]

**Pull request merged**